fish-rust/src/curses.rs

```diff
@@ -0,0 +1,300 @@
+    /// found. This is only called upon initialization of a [`Term`] instance in [`Term::new()`].
+    fn sys_lookup(&self) -> Option<CString> {
+        unsafe {
+            const NULL: *const i8 = core::ptr::null();
+            match sys::tgetstr(self.code.as_ptr(), core::ptr::null_mut()) {
+                NULL => None,
+                // termcap spec says nul is not allowed in terminal sequences and must be encoded;
+                // so the terminating NUL is the end of the string.
+                result => Some(CStr::from_ptr(result).to_owned()),
+            }
+        }
+    }
+}
+
+impl Capability for StringCap {
+    type Result = Option<Arc<CString>>;
+
+    /// We prepopulate all string capabilities at startup to never need to resize the vector, so we
+    /// never have to use [`StringCap::sys_lookup()`] here.
+    fn lookup(&self, term: &Term) -> Self::Result {
+        let idx = self.idx as u8 as usize;
+        term.strings[idx].borrow().clone()
+    }
+}
+
+impl Capability for Number {
+    type Result = Option<i32>;
+
+    fn lookup(&self, _: &Term) -> Self::Result {
+        unsafe {
+            match tgetnum(self.0.as_ptr()) {
+                -1 => None,
+                n => Some(n),
+            }
+        }
+    }
+}
+
+impl Capability for Flag {
+    type Result = bool;
+
+    fn lookup(&self, _: &Term) -> Self::Result {
+        unsafe { tgetflag(self.0.as_ptr()) != 0 }
+    }
+}
+
+impl StringCapIdx {
+    const fn idx(&self) -> usize {
+        *self as u8 as usize
+    }
+}
+
+/// Calls the curses `setupterm()` function with the provided `$TERM` value `term` (or a null
+/// pointer in case `term` is null) for the file descriptor `fd`.
+///
+/// Note that the `errret` parameter is provided to the function, meaning curses will not write
+/// error output to stderr in case of failure.
+///
+/// Any existing references from `curses::term()` will be invalidated by this call!
+///
+pub fn setup(term: Option<&CStr>, fd: i32) -> bool {
+    let result = unsafe {
+        // If cur_term is already initialized for a different $TERM value, calling setupterm() again
+        // will leak memory. Call reset() first to free previously allocated resources.
+        reset();
+
+        let mut err = 0;
+        if let Some(term) = term {
+            sys::setupterm(term.as_ptr(), fd, &mut err)
+        } else {
+            sys::setupterm(core::ptr::null(), fd, &mut err)
+        }
+    };
+
+    unsafe {
+        if result == sys::OK {
+            TERM = Some(Term::new());
+            true
+        } else {
+            TERM = None;
+            false
+        }
+    }
+}
+
+/// Whether or not the curses library has been initialized.
+pub fn is_initialized() -> bool {
+    unsafe { !sys::cur_term.is_null() }
+}
+
+/// Resets the curses `cur_term` TERMINAL pointer. Any previous term objects are invalidated!
+pub unsafe fn reset() {
```

Comment:
This should not be `pub`; it needs to be synchronized with `setupterm`.

Reply:
I'm not sure what you mean by "synchronized with `setupterm`", as `del_curterm` is a valid call after `setupterm()` and we set both `TERM` to `None` as `cur_term` gets nulled out. Obviously if we end up adding a mutex then it'll need to use the mutex, though.

(It needs to be public because it's the equivalent of `env_cleanup()` and gets called by asan at exit.)